src/utils/deferNetworkRequests.ts

```diff
@@ -0,0 +1,38 @@
+/**
+ * Intercepts and defers any requests on the page
+ * until the Service Worker instance is ready.
+ */
+export function deferNetworkRequests(
+  workerReady: Promise<ServiceWorkerRegistration | null>,
+) {
+  // Defer `XMLHttpRequest` until the Service Worker is ready.
+  const originalXhrSend = window.XMLHttpRequest.prototype.send
+  const restoreXhrSend = function (
+    this: XMLHttpRequest,
+    ...args: Parameters<XMLHttpRequest['send']>
+  ) {
+    window.XMLHttpRequest.prototype.send = originalXhrSend
+    this.send(...args)
+  }
+
+  window.XMLHttpRequest.prototype.send = function (
+    ...args: Parameters<XMLHttpRequest['send']>
+  ) {
+    workerReady
+      .then(() => restoreXhrSend.call(this, ...args))
+      .catch(() => restoreXhrSend.call(this, ...args))
+  }
+
+  // Defer `fetch` requests until the Service Worker is ready.
+  const originalFetch = window.fetch
+  const restoreFetch = (...args: Parameters<typeof window.fetch>) => {
+    window.fetch = originalFetch
+    return window.fetch(...args)
+  }
+
+  window.fetch = async (...args) => {
```

Comment:
`async` isn't necessary here?

Reply:
Actually, I think I will use `@open-draft/until` to handle both resolved and rejected Promise state. That way I don't have to repeat the logic in `.then` and `.catch`. 

That won't work for `XMLHttpRequest`, though, since it cannot be async.